admin/libs/firebase_util.js

```diff
@@ -0,0 +1,213 @@
+      const data = snapshot.val();
+      callback(data);
+    });
+  },
+
+  /**
+   * Saves a new bot mapping for an agent to the datastore.
+   *
+   * @param {str} fullAgentId The brand/ID/agent/ID
+   * @param {object} formObject The form key/value pairs
+   * @param {function} callback
+   * Callback method for after the method is complete.
+   */
+  saveBotInfo: (fullAgentId, formObject, callback) => {
+    const dfProjectId = formObject.projectId;
+    const dfServiceAccountKey = formObject.dfServiceAccountKey;
+
+    // remove brands from agent ID
+    const agentId = fullAgentId.substring(fullAgentId.lastIndexOf('/') + 1);
+    let csatTriggerWindow = formObject.csatTriggerWindow;
+
+    if (csatTriggerWindow.length == 0) {
+      csatTriggerWindow = 0;
+    }
+
+    const db = firebase.database();
+    const agentRef = db.ref('admin/agents').child(agentId);
+
+    const botInfo = {
+      dfServiceAccountKey: dfServiceAccountKey,
+      agentId: fullAgentId,
+      csatTriggerWindow: csatTriggerWindow,
+      projectId: dfProjectId,
+    };
+    // update botInfo in database
+    agentRef.set(botInfo, (err) => {
+      if (err) {
+        console.error(err);
+      } else {
+        if (callback) {
+          callback();
+        }
+      }
+    });
+  },
+
+  /**
+   * getConversationsList - gets list of conversations associated with provider
+   * @return {Array} List of conversations
+   */
+  getConversationsList: () => {
+    return new Promise((resolve, reject) => {
+      const db = firebase.database();
+      const conversationsRef = db.ref(`conversations/`);
+      conversationsRef.once('value', (snapshot) => {
+        const data = snapshot.val();
+        if (!data) {
+          resolve(conversations);
+        } else {
+          // search for conversations within agent, appending to list of
+          // conversations
+          const conversations = [];
+          const agentIds = Object.keys(data);
+          for (let i = 0; i < agentIds.length; i++) {
+            const agentId = agentIds[i];
+            const convIds = Object.keys(data[agentId]);
+            for (let j = 0; j < convIds.length; j++) {
+              const id = convIds[j];
+              conversations.push({
+                conversationId: id,
+                displayName: data[agentId][id].name,
+                agentId: agentId,
+              });
+            }
+          }
+          resolve(conversations);
+        }
+      });
+    });
+  },
+  /**
+   * createMessageListener - Creates a message to listen for the latest live
+   * agent messages and respond
+   * @param {String} agentId The Agent ID
+   * @param  {string} conversationId The Conversation ID
+   * @param  {function} callback Callback to run with message details
+   */
+  createMessageListener: (agentId, conversationId, callback) => {
+    const db = firebase.database();
+    const conversationsRef = db
+        .ref(`conversations/${agentId}/${conversationId}/messages`);
```

Comment:
This seems out of order, shouldn't this be conversations/${conversationId}/${agentId}...?

Reply:
I put agent ID first, because conversations are under each specific agent, and different agents get different conversationIds, even if they're the same person, from what I've seen